homeassistant/components/whirlpool/sensor.py

```diff
@@ -90,6 +92,26 @@ def washer_state(washer: WasherDryer) -> str | None:
     return MACHINE_STATE.get(machine_state, STATE_UNKNOWN)
 
 
+def completion_time(washer: WasherDryerClass) -> datetime | None:
+    """Calculate the time stamp for completion."""
+    machine_state = washer.washdry.get_machine_state()
+    if (
+        machine_state.value in {MachineState.Complete.value, MachineState.Standby.value}
+        and washer.running
+    ):
+        washer.running = False
+        washer.timestamp = utcnow()
```

Comment:
Can you help me understand the problem this is trying to solve by setting state on the entity? I really liked before how it was just taking in the `WasherDryer` and setting state on the entity, which seems like an unexpected side effect of a value fn

Reply:
I really liked how simple it used to be too :) 
It needs to keep the timestamp persistent at the end of a cycle and only update when running (aka NOT MachineState.Complete or MachineState.Standby).  this value function has no memory (that I'm aware of).   
I'm totally open to alternative methods of keeping the two variables: .running and .timestamp persistent.

Reply:
I added the machine_state variable rather than hitting  washer.washdry.get_machine_state() multiple times

Reply:
I added the machine_state variable rather than hitting  washer.washdry.get_machine_state() multiple times

Reply:
I was expecting something like this:
```
if washer.washdry.get_machine_state() is MachineState.RunningMainCycle:
    return utcnow() + timedelta(seconds=int(washer.washdry.get_attribute("Cavity_TimeStatusEstTimeRemaining")

return None
```

I don't get why this needs to add a setter and store state on the entity.

Reply:
If I do it this way, the entity shows "unknown" when the washer is not running.

The way it is implemented, it counts down until the cycle ends, then returns the timestamp for when they cycle ended (not None) from the stored timestamp.  
It then shows how long "until it ends", then when it ends, it shows "how long since it ended". 

I originally had it like your example with return utcnow() instead of return None, but that ends up updating the end time when the API receives a state change message - I needed to store utcnow() at the state change.

Am I way off base here?